### DISCUSSION TITLE
Energymanagement of hybrid heating system with PV

### DISCUSSION THREAD

**Discussion by Nxtway** [emsesp/EMS-ESP32#1550] [Q&A]
I want to improve the energy management  of the following hybrid installation:
- hybrid heating system (Buderus GBH212 gas boiler + WLW196i-6 heatpump + HM200 hybrid manager + RC310 thermostat)
- PV plant with 9 kWp without energy storage

In general I use the "cost optimized" hybrid strategy of the HM200. With 11ct gas and 37ct electricity costs the HM200 choose the heatpump on +2°C and higher and the gas boiler below +2°C. 

Now I use the electricity of my PV to feet the heatpump also on temperatures lower than +2°C.  Therefore I implement the following energymanagement strategy in node red and inject the parameter into the system with ems-esp commands:
![image](https://github.com/emsesp/EMS-ESP32/assets/82314801/596fcef7-265d-4c91-b8b7-b91aeee94990)

The result of the last 2 days is:
![image](https://github.com/emsesp/EMS-ESP32/assets/82314801/d1f05ca1-0c2e-4b3a-b9a3-e928ee422599)

**Now my question to you:** 
Has anybody an easier/better strategy ? Or other ideas to implement an energymanagement for a hybrid system ?

**Comment by Nxtway** [comment]
(How) can I use the SG-Ready functionality of the Buderus HM200 hybrid manager ? 
The SG-Ready functionality isn't visible in ems-esp and the only indicator in the RC310 is the display of the current mode in:
Servicemenu>>Diagnose>>Monitorwerte>>Wärmepume>Eingänge
In this Menu the SGReady mode 2 is displayed.
Can I switch HM200 input connector I4 via the ems-esp software ?
If not possible, then an alternativ might be connect the HM200 input I4 via an output of the ems-esp.
Therefore I found two discussions, but I have no idea what external hardware for a relay output I should buy?
[https://github.com/emsesp/EMS-ESP32/issues/1375](url)
[https://github.com/emsesp/EMS-ESP32/discussions/1270](url)
In the ems-esp documentation I found only examples for DS18** temperature sensors, but not for digital in- and outputs (espacially relay output).
Because of additional hardware and no cable wiring I prefer the software solution, if possible

**Reply by proddy** [reply]
I don't know what SG-Ready is but in EMS-ESP you can create an analog sensor to send outputs via a GPIO, like a relay.

**Reply by Nxtway** [reply]
Sorry, but I have really no idea how to use as DS1822 Temperature sensor as digital (relay) output ?
Exists there any special DS1822 electronic component with digital output functionality ?

**Reply by proddy** [reply]
I think I misunderstood your question; apologies. Forget my reply.

**Reply by MichaelDvP** [reply]
I also don't know what SGReady is. 
If you mean inputs like in #600, they are not implemented for HM200. If HM200 has the same features and uses the same telegram we can implement it. 
To connect a external relais like n #1375 you have to configure a gpio as digital out in the analog sensor section,

**Reply by Nxtway** [reply]
I didn't find any option in the RC310 Menu to change the logic of the HM200.2 pins I1 and I4. From the SG-Ready standard point of view I imagine that's also not wanted.
Also I read the [https://github.com/emsesp/EMS-ESP32/issues/1375](url). At the end wrote BBQKees: "Furthermore I do not support switching relays from a BBQKees Gateway." 
Details in [https://bbqkees-electronics.nl/wiki/gateway/using-gateway-gpio.html](url) "Although the EMS-ESP firmware has a feature to add external sensors and relays, BBQKees Gateways are not designed for this purpose. "
This is probably the end of the journey

**Comment by Nxtway** [comment]
Hello,
I collect several information in the meantime.
SG-Ready is described in the HM200.2 manual (de, fr, it, **nl-BE**): [https://buderus-de-de.boschtt-documents.com/download/pdf/file/6721864357.pdf](url)
(you have to copy&paste this link into your browser, the link function doesn't work in this form)
See chapter 2.5.5

I want to read (and wish to write) the SG-Ready mode.
![image](https://github.com/emsesp/EMS-ESP32/assets/82314801/ca9fdd90-c9f0-4e12-ae38-cc20610a911d)

I'm now able to read the mode via:
![image](https://github.com/emsesp/EMS-ESP32/assets/82314801/7443ad94-ca5c-47ca-860d-2bff69a31133)
I checked all 4 states using the HM200.2 inputs I1 and I4: 0=Mode1 | 1=Mode2 | 2=Mode3 | 3=Mode4
The I4 is closed and on the screenshot is shown the value 2, meaning Mode3 - heatpump preferd

1. Question: _Is the mode also be writable as command from ems-esp? The advantage will be that's not needed to use the hardware input I4 to toggle between standard mode and heatpump prefered._ I imagine it's not possible ?
2. Question: If it's not possible to set this function via software I have to do this by hardware. Therefore I need to open/close the HM200.2 I4 input with the GPIO function of ems-esp. If have no idea how to wire a relay to the ems-esp gpio port and at the end of  [https://github.com/emsesp/EMS-ESP32/issues/1375](https://github.com/emsesp/EMS-ESP32/discussions/url) BBQKees wrote: "Furthermore I do not support switching relays from a BBQKees Gateway." 

This is probably the end of the journey of writing the SG Ready Mode 1..4 via software or hardware (BBQKees Gateway S3 on the Buderus HM200.2). But reading of the current Mode SG Ready Mode is possible (see screenshots).

**Reply by MichaelDvP** [reply]
Set the custom `SG_Mode_Ready` to type UINT and writable, then you can go to dashboard and write a new value (also via api/mqtt/console).
I don't know if you can set the inputs to `invert`. If so, you need to test what telegram/offset this is and we can add it.

**Reply by Nxtway** [reply]
I tried it, but the heatpump ignored it, logfile attached: [SG_Ready_try_mode_set.txt](https://github.com/emsesp/EMS-ESP32/files/14000557/SG_Ready_try_mode_set.txt) search for "0x099A"

In discussion [https://github.com/emsesp/EMS-ESP32/issues/600](url) are some pictures to change the logic of the heatpump.
This seems to be another model. In my RC310 I didn't find any menu to change the logic for the HM200.2

**Reply by MichaelDvP** [reply]
Yes, 0x99A seems to be a monitor telegram, not writeable. (In `T 100: [emsesp] heatpump(0x53) -B-> All(0x00), ?(0x099A)`it is broadcasted, setting telegrams normally only broadcast a change, never the complete telegram).
You have to find the setting. Log and change the setting on the thermostat, you should see a message `thermostat(0x10) -W-> Heatpump(0x53), ?(telegram), data: xx, offset yy` with a single byte of data. This is the writable telegram.

**Reply by HWE70** [reply]
@Nxtway where you able to fix the issue?
